// @langchain/langgraph/dist/errors.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidUpdateError = exports.EmptyChannelError = exports.EmptyInputError = exports.isGraphInterrupt = exports.NodeInterrupt = exports.GraphInterrupt = exports.GraphValueError = exports.GraphRecursionError = void 0;
class GraphRecursionError extends Error {
    constructor(message) {
        super(message);
        this.name = "GraphRecursionError";
    }
    static get unminifiable_name() {
        return "GraphRecursionError";
    }
}
exports.GraphRecursionError = GraphRecursionError;
class GraphValueError extends Error {
    constructor(message) {
        super(message);
        this.name = "GraphValueError";
    }
    static get unminifiable_name() {
        return "GraphValueError";
    }
}
exports.GraphValueError = GraphValueError;
class GraphInterrupt extends Error {
    constructor(interrupts = []) {
        super(JSON.stringify(interrupts, null, 2));
        Object.defineProperty(this, "interrupts", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = "GraphInterrupt";
        this.interrupts = interrupts;
    }
    static get unminifiable_name() {
        return "GraphInterrupt";
    }
}
exports.GraphInterrupt = GraphInterrupt;
/** Raised by a node to interrupt execution. */
class NodeInterrupt extends GraphInterrupt {
    constructor(message) {
        super([
            {
                value: message,
                when: "during",
            },
        ]);
        this.name = "NodeInterrupt";
    }
    static get unminifiable_name() {
        return "NodeInterrupt";
    }
}
exports.NodeInterrupt = NodeInterrupt;
function isGraphInterrupt(e) {
    return (e !== undefined &&
        [
            GraphInterrupt.unminifiable_name,
            NodeInterrupt.unminifiable_name,
        ].includes(e.name));
}
exports.isGraphInterrupt = isGraphInterrupt;
class EmptyInputError extends Error {
    constructor(message) {
        super(message);
        this.name = "EmptyInputError";
    }
    static get unminifiable_name() {
        return "EmptyInputError";
    }
}
exports.EmptyInputError = EmptyInputError;
class EmptyChannelError extends Error {
    constructor(message) {
        super(message);
        this.name = "EmptyChannelError";
    }
    static get unminifiable_name() {
        return "EmptyChannelError";
    }
}
exports.EmptyChannelError = EmptyChannelError;
class InvalidUpdateError extends Error {
    constructor(message) {
        super(message);
        this.name = "InvalidUpdateError";
    }
    static get unminifiable_name() {
        return "InvalidUpdateError";
    }
}
exports.InvalidUpdateError = InvalidUpdateError;
